backend/fms_core/services/sample.py

```diff
@@ -546,9 +546,49 @@ def pool_samples(process: Process,
 
     return sample_destination, errors, warnings
 
+def test_and_update_volume_ratio(samples: list[dict]):
+    """
+    Tests the volume ratio of the samples in the pool and updates them if necessary.
+    If all samples have volume_ratio defined, it checks that they are all defined and equal.
+    If all samples do not have volume_ratio defined, it sets the volume_ratio to 1/n for each sample.
+    If some samples have volume_ratio defined and others do not, it raises an error.
+    If the volume ratios do not add up to 1, it raises an error.
+
+    Args:
+        samples: A list of sample dictionaries, each containing at least key "alias" and optionally "volume_ratio".
+
+    Returns:
+        A list of errors and the updated samples with volume_ratio.
+    """
+
+    errors: list[str] = []
+
+    equal_volume_ratio = True
+    for sample in samples:
+        if sample.get("volume_ratio", None) is not None:
+            equal_volume_ratio = False
+
+    for sample in samples:
+        if not equal_volume_ratio and sample.get("volume_ratio", None) is None:
+            errors.append(f"Volume ratio for sample {sample['alias']} is missing but all samples must either define volume ratio or not simultaneously.")
+    if errors:
+        return errors
+
+    exp = Decimal(f'1E-{DerivedBySample.VOLUME_RATIO_DECIMAL_PLACES}')
+    if equal_volume_ratio:
+        for sample in samples:
+            sample["volume_ratio"] = (Decimal(1) / Decimal(len(samples))).quantize(exp)
+    else:
+        # Verify that the ratio add up to 1
+        total_volume_ratio = Decimal(sum(Decimal(sample["volume_ratio"]).quantize(exp) for sample in samples)).quantize(exp)
```

Comment:
You loop around samples a lot ... try to store the information required here during the initial loop. If there is missing ratios, the value is n * 1/n ... Also by doing the rounding so each sample ratio you increase your risk of having rounding errors i think. For example, you have 3 sample without ratio you allocate the value 1/3 to them the rounding will add 0.333...3 three times which will not add up to 1 at the 15 decimal precision, but if you add 1/3 at the full precision 3 times then round them up to the 15th decimal they will round up to 1.